tensorflow/core/common_runtime/gpu/gpu_util.cc

```diff
@@ -378,10 +397,11 @@ void GPUUtil::CopyCPUTensorToGPU(const Tensor* cpu_tensor,
     }
   }
 
+  if (merge_host_to_device_stream) done(OkStatus());
```

Comment:
Is it possible that something bad happens when the copy is actually performed (i.e, there is an error when doing copy)? How/when will that error be returned?

Reply:
Yes, it's possible. Just like in the original implementation, the error will cause the subsequent cuda event poll in `dev_info->event_mgr->ThenExecute()` to return error ([here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/device/device_event_mgr.cc#L220-L224)).
In fact, we can put the `if (!recv_host_to_device_stream->ok())` ahead like the `done` function as well, like this:

```suggestion
  if (merge_host_to_device_stream) {
    if (!recv_host_to_device_stream->ok()) {
      LOG(FATAL) << "CPU->GPU Memcpy failed";
    }
    done(OkStatus());
  }
  dev_info->event_mgr->ThenExecute(
      recv_host_to_device_stream,
      [recv_host_to_device_stream, done, input_ref, do_staging, staging_buffer,
       host_memory_allocator, merge_host_to_device_stream]() {
        if (do_staging) {
          host_memory_allocator->DeallocateRaw(staging_buffer);
        } else {
          input_ref.Unref();
        }
        if (!merge_host_to_device_stream) {
          if (!recv_host_to_device_stream->ok()) {
            LOG(FATAL) << "CPU->GPU Memcpy failed";
          }
          done(OkStatus());
        }
      });
```
This is because `recv_host_to_device_stream->ok()` only checks the previous **launch** of `recv_host_to_device_stream->ThenMemcpy()` ([here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/stream_executor/stream.cc#L1767)). It doesn't check the execution status of doing copy at all.

Reply:
(sorry, it's been a while) one more question

> It doesn't check the execution status of doing copy at all.

The status check `if (!recv_host_to_device_stream->ok()){...}` is in the callback in `dev_info->event_mgr->ThenExecute(stream, callback)` - that is, the status check is not performed until all operations (including the data copy here) before the event (added during callback enqueue) on the stream have completed, but not just launch.

So originally if there is an error caused by data transfer, the callback will check the stream status and return. With stream-merging, there is no such callback, an error caused by data transfer will not be detected immediately, but rely on some other checks to reveal it, maybe at a slightly later time.

Am I understanding it correctly? is it a problem?

Reply:
I didn't find anywhere in the code that `PollEvents(stream)` can set `recv_host_to_device_stream->status_` to not-ok. When the event poll fails (maybe caused by a data copy error as you assumed), the code simply goes to [`LOG(FATAL)`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/device/device_event_mgr.cc#L220-L224), instead of updating the stream status via `recv_host_to_device_stream->CheckStatus(s)`. So, `recv_host_to_device_stream->ok()` will always return true.

And with stream-merging, there is still a callback to deal with the `if (do_staging) {...} else {...}` logic, so event poll still happens, and `LOG(FATAL)` is still there.

Please correct me if something is wrong in my understanding.

Reply:
I think `PollEvent()` is intended to only check event status, instead of updating stream status. Stream status is updated in other places, via e.g., `Stream::CheckStatus()` ,`Stream::RefreshStatus()`. Unless stream status can be always implied by event status, we can not rely on `PollEvent()` to tell the stream status.

When we do
```
  if (merge_host_to_device_stream) {
    if (!recv_host_to_device_stream->ok()) {
      LOG(FATAL) << "CPU->GPU Memcpy failed";
    }
    done(OkStatus());
  }
```
the newly launched data transfer could be pending, and `if (!recv_host_to_device_stream->ok()){}` only checks the  stream status for the completed work that comes before this data transfer, and then performs `done(OkStatus());` without knowing the actual execution status of the data transfer. Is it ok to fire the callback `done(status)` before we know the actual data transfer status, in the case of stream merging?

Reply:
I see. I think we are confusing two definitions of 'stream status' here. Let's make it clear:

1. CUDA-level status. In the design of CUDA, if one operation (cuda kernel, memcopy, ...) fails, all subsequent operations, events, and kernel launches in the stream will return error as well. So, here if the data transfer failed for some reason, the subsequent `PollEvent` (which will call `cuEventQuery` API repeatedly internally) will return error, too.
2. TF-level status, which is a member variable of class `Stream`, can be set by `Stream::CheckStatus()` and `Stream::RefreshStatus()`, and queried with `Stream::ok()`.

So, based on these two definitions, let me try to answer your questions:

> Unless stream status can be always implied by event status, we can not rely on PollEvent() to tell the stream status.

Yes, CUDA-level stream status can be always implied by event status.

> Is it ok to fire the callback done(status) before we know the actual data transfer status, in the case of stream merging?

If 'data transfer status' refers to the TF-level status, what really happens is, you never know the actual data transfer status, no matter with or without stream merging. Because as far as I know, there is nowhere TF-level stream status is updated in the whole data transfer procedure, so `recv_host_to_device_stream->ok()` will always return true. On the other hand, if 'data transfer status' refers to the CUDA-level status, firing the callback `done(status)` before we know the actual data transfer status looks like a kind of **prefetching** to schedule subsequent ops. I admit it seems a bit risky, but I can't think of any substantial harm either. I suppose it's fine because the later event poll will report error anyway if the data transfer failed.

Reply:
Thanks for the explanation. I agree that it will error out eventually if a data transfer has an error.

> firing the callback done(status) before we know the actual data transfer status looks like a kind of prefetching to schedule subsequent ops.

Depending on the operations in the callback, firing the callback earlier before data transfer is done could be problematic. Take [this](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/runtime_fallback/kernel/tensor_util.h#L100-L112) for an example. The purpose of the callback is to notify that the transfer is done
```
   status = s;
   n.Notify();
```
so that the caller can start to use the copy
```
  result.emplace(std::move(dst));
```
If we invoke the callback earlier before the transfer is done, then this logic seems problematic.

Another example [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device.cc#L996-L1004).

Reply:
Yes, and that's why we move the callback ahead only **in stream-merging mode** and only **in HtoD copy** (not in DtoH). Only in this way it's safe. Because all subsequent operations to the transferred destination data will be queued in the **same** stream as the copy, so there is a CUDA-promised dependency. Think about sequential TF ops scheduling doesn't care about GPU execution status, it's the same thing.

If either of these two conditions is not met, it will be problematic as you said. 1) If it's not in stream-merging mode, we cannot move the callback ahead, because subsequent operations won't check the copy-stream. 2) If it's a DtoH copy, we cannot move the callback ahead, because subsequent operations will be pure CPU behaviors and are out of CUDA's dependency promision.

Reply:
Thanks, that's the important point to make! Can you help to add a comment to the code to explain this briefly?

Reply:
Add some comments here: https://github.com/tensorflow/tensorflow/pull/61632/files#diff-79e98c37046740086c03187fc10acc2c801899d8fa152fc7f94aae2dddf9a149R480-R491